In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df = pd.read_excel(r"C:\Users\toshu\OneDrive\Desktop\bootcamp\13. EDA_Cars_Data.xlsx")

In [3]:
df.head()

INDEX        INCOME MARITAL STATUS SEX    EDUCATION          JOB  \
0      1  125301.24250             No   F    Bachelors  Blue Collar   
1      2   50815.44531             No   M  High School          NaN   
2      3   62977.82416            NaN   F    Bachelors     Clerical   
3      4   77099.96624             No   F          NaN       Lawyer   
4      5  130794.57420             No   M  High School          NaN   

   TRAVEL TIME         USE  MILES CLOCKED     CAR TYPE CAR AGE   CITY  \
0    45.703013  Commercial        17430.0   Sports Car       7  Texas   
1    20.591628     Private        18930.0      Minivan       1  Texas   
2    33.639949     Private            NaN          SUV       ?  Texas   
3    15.415676         NaN        18300.0   Sports Car      11  Texas   
4          NaN  Commercial        28340.0  Panel Truck      10  Texas   

   POSTAL CODE  
0      42420.0  
1      42420.0  
2      42420.0  
3      42420.0  
4      42420.0

Observation
there are null values in the dataset
car age might have wrong entriues like'?'

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   INDEX           303 non-null    int64  
 1   INCOME          260 non-null    float64
 2   MARITAL STATUS  275 non-null    object 
 3   SEX             297 non-null    object 
 4   EDUCATION       259 non-null    object 
 5   JOB             257 non-null    object 
 6   TRAVEL TIME     262 non-null    float64
 7   USE             250 non-null    object 
 8   MILES CLOCKED   278 non-null    float64
 9   CAR TYPE        293 non-null    object 
 10  CAR AGE         283 non-null    object 
 11  CITY            297 non-null    object 
 12  POSTAL CODE     300 non-null    float64
dtypes: float64(4), int64(1), object(8)
memory usage: 30.9+ KB


there are 1 column of int datatype
there are 4 columns of float datatype
the remaining are object

In [6]:
df.shape

(303, 13)

There are 303 rows in our dataset
there are 13 columns in our dataset

In [7]:
df.dtypes

INDEX               int64
INCOME            float64
MARITAL STATUS     object
SEX                object
EDUCATION          object
JOB                object
TRAVEL TIME       float64
USE                object
MILES CLOCKED     float64
CAR TYPE           object
CAR AGE            object
CITY               object
POSTAL CODE       float64
dtype: object

observation
features are having null values
the data type of car age should be integer but it is appearing as object

In [8]:
df.describe()

INDEX         INCOME  TRAVEL TIME  MILES CLOCKED   POSTAL CODE
count  303.000000     260.000000   262.000000     278.000000    300.000000
mean   139.640264   50025.162170    34.282098   13591.978417  50712.196667
std     85.178422   41186.807914    14.910178    7167.328655  24141.029290
min      1.000000       0.000000     5.000000    1500.000000  11435.000000
25%     62.500000   20452.885023    24.449874    7900.000000  42420.000000
50%    138.000000   44571.590870    33.564757   12065.000000  47150.000000
75%    213.500000   66485.761387    43.907339   18240.000000  61701.000000
max    289.000000  204667.589700    83.617643   38000.000000  90049.000000

observation
in income we have min as 0 so we need to check the validity of data with people having zero income and having vehicles

In [9]:
df[df['INCOME'] == 0]

INDEX  INCOME MARITAL STATUS  SEX    EDUCATION         JOB  TRAVEL TIME  \
23      24     0.0            Yes    F  High School  Home Maker    24.520001   
24      25     0.0            NaN    M  High School     Student    13.662860   
34      35     0.0             No    M  High School         NaN    46.932840   
40      41     0.0             No    M  High School  Home Maker    36.220252   
66      67     0.0             No    M  High School     Student    32.525872   
94      81     0.0             No    F  High School  Home Maker    40.956033   
95      82     0.0            Yes    F          NaN  Home Maker    10.700788   
103     90     0.0             No    F  High School     Student    18.462764   
125    112     0.0            Yes    F          NaN  Home Maker    12.025304   
140    127     0.0            Yes    M  High School     Student    35.269037   
142    129     0.0            Yes    F  High School     Student    47.996642   
143    130     0.0            Yes    F  High School  Home Maker    28.214852   
144    131     0.0            NaN    M  High School     Student    28.859550   
173    160     0.0            Yes    F  High School  Home Maker    33.179012   
201    188     0.0             No    F      Masters  Home Maker          NaN   
207    194     0.0             No    M  High School         NaN    23.748839   
251    238     0.0            Yes    F  High School     Student          NaN   
255    242     0.0             No    M          NaN  Home Maker    22.322261   
259    246     0.0             No    M          NaN  Home Maker    29.215601   
265    252     0.0             No    F  High School     Student    22.845634   
273    260     0.0             No    M          NaN     Student    34.384168   
285    272     0.0            Yes    F    Bachelors     Student    50.484397   
288    275     0.0             No    F      Masters  Home Maker    43.018894   
297    284     0.0             No  NaN  High School     Student    15.379303   
302    289     0.0             No    M    Bachelors     Student    11.657483   

            USE  MILES CLOCKED    CAR TYPE CAR AGE           CITY  POSTAL CODE  
23      Private        11740.0         SUV       1          Texas      42420.0  
24   Commercial        15090.0     Minivan       5   Philadelphia      84057.0  
34   Commercial         5700.0      Pickup       1    Los Angeles      90049.0  
40          NaN        16510.0     Minivan       1    Los Angeles      90049.0  
66      Private            NaN      Pickup       1  New York City      12180.0  
94      Private        14240.0         SUV       1  New York City      12180.0  
95      Private         8560.0  Sports Car      14  New York City      12180.0  
103         NaN         1500.0         SUV       1  San Francisco      55106.0  
125         NaN         6500.0  Sports Car     NaN        Houston      61701.0  
140  Commercial        11610.0      Pickup       1        Houston      61701.0  
142     Private         7740.0  Sports Car       1        Houston      61701.0  
143     Private         5100.0     Minivan       9        Houston      61701.0  
144     Private         9240.0      Pickup       1        Houston      61701.0  
173     Private         5200.0         SUV       6        Seattle      78207.0  
201     Private        12790.0     Minivan      18       Columbia      44123.0  
207  Commercial         5800.0  Sports Car       1       Columbia      44123.0  
251         NaN        16010.0         SUV       1     Washington      11435.0  
255     Private            NaN         NaN      10     Washington      11435.0  
259     Private        17700.0         SUV      12     Washington      11435.0  
265     Private         6600.0         SUV       1      Las Vegas      44412.0  
273         NaN         6000.0         SUV       7      Las Vegas      44412.0  
285     Private         9270.0         SUV      10            NaN      44412.0  
288     Private         5600.0  Sports Car      20      Las V

observations
it seems that most of the entries with income as 0 are either students or home makers so entries might be genuine

# Data cleaning

## 1. correcting data type

In [10]:
#changing the postal code datatype into object
df['POSTAL CODE'] = df['POSTAL CODE'].astype('object')

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   INDEX           303 non-null    int64  
 1   INCOME          260 non-null    float64
 2   MARITAL STATUS  275 non-null    object 
 3   SEX             297 non-null    object 
 4   EDUCATION       259 non-null    object 
 5   JOB             257 non-null    object 
 6   TRAVEL TIME     262 non-null    float64
 7   USE             250 non-null    object 
 8   MILES CLOCKED   278 non-null    float64
 9   CAR TYPE        293 non-null    object 
 10  CAR AGE         283 non-null    object 
 11  CITY            297 non-null    object 
 12  POSTAL CODE     300 non-null    object 
dtypes: float64(3), int64(1), object(9)
memory usage: 30.9+ KB


# 2. duplicate rows

In [12]:
#checking the number of duplicated rows
df.duplicated().sum()

13

In [13]:
#for checking the duplicated values
df[df.duplicated() == True]

INDEX       INCOME MARITAL STATUS SEX    EDUCATION          JOB  \
69     29  64013.81632            Yes   M  High School  Blue Collar   
70     29  64013.81632            Yes   M  High School  Blue Collar   
71     29  64013.81632            Yes   M  High School  Blue Collar   
72     29  64013.81632            Yes   M  High School  Blue Collar   
73     29  64013.81632            Yes   M  High School  Blue Collar   
74     29  64013.81632            Yes   M  High School  Blue Collar   
75     29  64013.81632            Yes   M  High School  Blue Collar   
76     29  64013.81632            Yes   M  High School  Blue Collar   
77     29  64013.81632            Yes   M  High School  Blue Collar   
78     29  64013.81632            Yes   M  High School  Blue Collar   
79     29  64013.81632            Yes   M  High School  Blue Collar   
80     29  64013.81632            Yes   M  High School  Blue Collar   
81     29  64013.81632            Yes   M  High School  Blue Collar   

    TRAVEL TIME         USE  MILES CLOCKED CAR TYPE CAR AGE         CITY  \
69    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
70    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
71    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
72    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
73    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
74    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
75    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
76    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
77    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
78    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
79    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
80    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   
81    32.717234  Commercial         7900.0   Pickup       5  Los Angeles   

   POSTAL CODE  
69     90049.0  
70     90049.0  
71     90049.0  
72     90049.0  
73     90049.0  
74     90049.0  
75     90049.0  
76     90049.0  
77     90049.0  
78     90049.0  
79     90049.0  
80     90049.0  
81     90049.0

In [14]:
#permanently removing the duplicated rows
df.drop_duplicates(inplace = True)

In [15]:
#checking number of duplicated rows
df.duplicated().sum()

0

## we can now see that we have no duplicated values now

In [16]:
df.shape

(290, 13)

## Observation after removing the duplicated values we now have 290 rows 

## 3. Invalid and wrong values/ entries

In [17]:
df['CAR AGE'].unique()

array([7, 1, '?', 11, 10, 5, nan, 6, 13, 4, 15, 3, 8, 14, 20, 16, 17, 12,
       9, 18, 19], dtype=object)

In [18]:
df['CAR AGE'].value_counts()

CAR AGE
1     97
10    21
8     21
6     19
7     16
11    15
9     12
5     12
4     11
12     7
14     6
20     6
3      5
16     4
13     4
18     4
15     3
17     3
?      3
19     1
Name: count, dtype: int64

In [19]:
df[df['CAR AGE'] == '?']

INDEX       INCOME MARITAL STATUS SEX    EDUCATION          JOB  \
2        3  62977.82416            NaN   F    Bachelors     Clerical   
82      29  64013.81632            Yes   M  High School  Blue Collar   
113    100          NaN             No   F      Masters          NaN   

     TRAVEL TIME         USE  MILES CLOCKED CAR TYPE CAR AGE           CITY  \
2      33.639949     Private            NaN      SUV       ?          Texas   
82     32.717234  Commercial         7900.0   Pickup       ?    Los Angeles   
113    42.376470  Commercial        24410.0   Pickup       ?  San Francisco   

    POSTAL CODE  
2       42420.0  
82      90049.0  
113     55106.0

In [20]:
df['CAR AGE'] = df['CAR AGE'].replace('?',np.nan)

In [21]:
df['CAR AGE'].unique()

array([ 7.,  1., nan, 11., 10.,  5.,  6., 13.,  4., 15.,  3.,  8., 14.,
       20., 16., 17., 12.,  9., 18., 19.])

In [22]:
df['CAR AGE'].value_counts()

CAR AGE
1.0     97
10.0    21
8.0     21
6.0     19
7.0     16
11.0    15
5.0     12
9.0     12
4.0     11
12.0     7
20.0     6
14.0     6
3.0      5
16.0     4
13.0     4
18.0     4
17.0     3
15.0     3
19.0     1
Name: count, dtype: int64

In [23]:
df[df['CAR AGE'] == '?']

Empty DataFrame
Columns: [INDEX, INCOME, MARITAL STATUS, SEX, EDUCATION, JOB, TRAVEL TIME, USE, MILES CLOCKED, CAR TYPE, CAR AGE, CITY, POSTAL CODE]
Index: []

### 4. missing values/null values 

In [24]:
df.isnull().sum()

INDEX              0
INCOME            43
MARITAL STATUS    28
SEX                6
EDUCATION         44
JOB               46
TRAVEL TIME       41
USE               53
MILES CLOCKED     25
CAR TYPE          10
CAR AGE           23
CITY               6
POSTAL CODE        3
dtype: int64

In [25]:
df[df.isnull().sum(axis = 1)>0]

INDEX        INCOME MARITAL STATUS  SEX    EDUCATION         JOB  \
1        2   50815.44531             No    M  High School         NaN   
2        3   62977.82416            NaN    F    Bachelors    Clerical   
3        4   77099.96624             No    F          NaN      Lawyer   
4        5  130794.57420             No    M  High School         NaN   
5        6           NaN            NaN    F  High School  Home Maker   
..     ...           ...            ...  ...          ...         ...   
295    282   52626.24977             No    F    Bachelors    Clerical   
297    284       0.00000             No  NaN  High School     Student   
299    286   18408.39545             No    M  High School    Clerical   
300    287           NaN             No    M  High School    Clerical   
301    288           NaN            Yes    F  High School         NaN   

     TRAVEL TIME         USE  MILES CLOCKED     CAR TYPE  CAR AGE       CITY  \
1      20.591628     Private        18930.0      Minivan      1.0      Texas   
2      33.639949     Private            NaN          SUV      NaN      Texas   
3      15.415676         NaN        18300.0   Sports Car     11.0      Texas   
4            NaN  Commercial        28340.0  Panel Truck     10.0      Texas   
5      48.360191         NaN         6000.0          SUV      5.0      Texas   
..           ...         ...            ...          ...      ...        ...   
295          NaN     Private        10300.0       Pickup      NaN  Las Vegas   
297    15.379303     Private         5800.0   Sports Car      6.0  Las Vegas   
299    42.619118     Private         7440.0          NaN      1.0  Las Vegas   
300    36.533624     Private         7660.0          SUV      NaN  Las Vegas   
301    60.884292     Private        19550.0          Van      1.0  Las Vegas   

    POSTAL CODE  
1       42420.0  
2       42420.0  
3       42420.0  
4       42420.0  
5       42420.0  
..          ...  
295     44412.0  
297     44412.0  
299         NaN  
300     44412.0  
301     44412.0  

[196 rows x 13 columns]

In [26]:
df[df.isnull().any(axis = 1)]

INDEX        INCOME MARITAL STATUS  SEX    EDUCATION         JOB  \
1        2   50815.44531             No    M  High School         NaN   
2        3   62977.82416            NaN    F    Bachelors    Clerical   
3        4   77099.96624             No    F          NaN      Lawyer   
4        5  130794.57420             No    M  High School         NaN   
5        6           NaN            NaN    F  High School  Home Maker   
..     ...           ...            ...  ...          ...         ...   
295    282   52626.24977             No    F    Bachelors    Clerical   
297    284       0.00000             No  NaN  High School     Student   
299    286   18408.39545             No    M  High School    Clerical   
300    287           NaN             No    M  High School    Clerical   
301    288           NaN            Yes    F  High School         NaN   

     TRAVEL TIME         USE  MILES CLOCKED     CAR TYPE  CAR AGE       CITY  \
1      20.591628     Private        18930.0      Minivan      1.0      Texas   
2      33.639949     Private            NaN          SUV      NaN      Texas   
3      15.415676         NaN        18300.0   Sports Car     11.0      Texas   
4            NaN  Commercial        28340.0  Panel Truck     10.0      Texas   
5      48.360191         NaN         6000.0          SUV      5.0      Texas   
..           ...         ...            ...          ...      ...        ...   
295          NaN     Private        10300.0       Pickup      NaN  Las Vegas   
297    15.379303     Private         5800.0   Sports Car      6.0  Las Vegas   
299    42.619118     Private         7440.0          NaN      1.0  Las Vegas   
300    36.533624     Private         7660.0          SUV      NaN  Las Vegas   
301    60.884292     Private        19550.0          Van      1.0  Las Vegas   

    POSTAL CODE  
1       42420.0  
2       42420.0  
3       42420.0  
4       42420.0  
5       42420.0  
..          ...  
295     44412.0  
297     44412.0  
299         NaN  
300     44412.0  
301     44412.0  

[196 rows x 13 columns]

In [27]:
#check for percentage of missing values
df.isnull().sum()/len(df)*100

INDEX              0.000000
INCOME            14.827586
MARITAL STATUS     9.655172
SEX                2.068966
EDUCATION         15.172414
JOB               15.862069
TRAVEL TIME       14.137931
USE               18.275862
MILES CLOCKED      8.620690
CAR TYPE           3.448276
CAR AGE            7.931034
CITY               2.068966
POSTAL CODE        1.034483
dtype: float64

if any of our column have more than 20 percent of null values then we either remove that column or we conduct a separate study for that column

In [ ]:
dor i in range(

find the outliers using the boxplot and find the iqr also

### np.where(df['income']>upperlimit, upperlimit,df['income']) 

normalization using the scipy libarary 
importing the stats.boxcox

In [ ]:
scaling is the method used to change the range of features of data